app.py

```diff
@@ -2,15 +2,68 @@
 import numpy as np
 from kmeans.parallel import ParallelKMeans
 from kmeans.sequential import SequentialKMeans
+import csv
+import os
+import pandas as pd
 
 np.random.seed(1234)
 
 K = 3
-N = 10000
 M = 2
 max_iter = 10
 
-X = np.random.rand(N, M)
+
+
+# Initialize MPI
+comm = MPI.COMM_WORLD
+rank = comm.Get_rank()
+size = comm.Get_size()
+
+csv_directory = './data' 
+
+import os
+import csv
+import numpy as np
+
+def read_csv_files_in_directory(directory):
+    data_list = []
+
+    for filename in os.listdir(directory):
+        if filename.endswith(".csv"):
+            file_path = os.path.join(directory, filename)
+
+            with open(file_path, 'r', newline='') as csv_file:
+                csv_reader = csv.reader(csv_file)
+                data = [row for row in csv_reader]
+
+            data_list.extend(data)
+
+    numpy_array = np.array(data_list)
+
+    return numpy_array.astype(np.float64)
+
+X = read_csv_files_in_directory(csv_directory)
+
+
+
+
+csv_files = [os.path.join(csv_directory, file) for file in os.listdir(csv_directory) if file.endswith(".csv")]
+
+total_row_count = 0
+for file_path in csv_files:
+    with open(file_path, 'r') as file:
+        total_row_count += sum(1 for line in file)
+
+rows_per_process = total_row_count // size
+
+start_row = rank * rows_per_process
+
+end_row = start_row + rows_per_process
+
+for file_path in csv_files:
+    df = pd.read_csv(file_path, skiprows=range(1, start_row), nrows=rows_per_process)
```

Comment:
No, this is not how it suppose to be done! You are reading every csv and load it to the memory turning this task somewhat similar to task 1. 

Do a calculation on how many rows to read in all csv files. Develop an algorithm that optimized for memory and IO or a simple function that generate a dict or something that says "from file x, starting from line y, read z rows" and and pass it to the processes.